### ISSUE TITLE
Infinite loop in TLSActor

### ISSUE THREAD

**Issue by maxcom** [akka/akka#29922]
We found that several threads in our application stuck in infinite loop in TLSActor. It happens without any active TCP connection (all connections was closed more than 10 minutes ago). Here is few thread dumps:

```
"application-akka.actor.default-dispatcher-11" #53 prio=5 os_prio=0 cpu=85937645.88ms elapsed=97848.06s tid=0x00007fa74c002000 nid=0x1a1b runnable  [0x00007fa77849c000]
   java.lang.Thread.State: RUNNABLE
        at akka.util.ByteString.isEmpty(ByteString.scala:758)
        at akka.util.ByteString$ByteString1C.$plus$plus(ByteString.scala:210)
        at akka.stream.impl.io.TLSActor$ChoppingBlock.putBack(TLSActor.scala:125)
        at akka.stream.impl.io.TLSActor.doUnwrap(TLSActor.scala:400)
        at akka.stream.impl.io.TLSActor.doInbound(TLSActor.scala:297)
        at akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1(TLSActor.scala:232)
        at akka.stream.impl.io.TLSActor$$Lambda$2467/0x0000000840ee7840.apply$mcV$sp(Unknown Source)
        at akka.stream.impl.Pump.pump(Transfer.scala:203)
        at akka.stream.impl.Pump.pump$(Transfer.scala:201)
        at akka.stream.impl.io.TLSActor.pump(TLSActor.scala:52)
        at akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse(ActorProcessor.scala:244)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
        at akka.stream.impl.SubReceive.apply(Transfer.scala:19)
        at akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse(FanOut.scala:242)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
        at akka.stream.impl.SubReceive.apply(Transfer.scala:19)
        at akka.stream.impl.SubReceive.apply(Transfer.scala:15)
        at scala.PartialFunction.applyOrElse(PartialFunction.scala:189)
        at scala.PartialFunction.applyOrElse$(PartialFunction.scala:188)
        at akka.stream.impl.SubReceive.applyOrElse(Transfer.scala:15)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:244)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:245)
        at akka.actor.Actor.aroundReceive(Actor.scala:537)
        at akka.actor.Actor.aroundReceive$(Actor.scala:535)
        at akka.stream.impl.io.TLSActor.aroundReceive(TLSActor.scala:52)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:577)
        at akka.actor.ActorCell.invoke(ActorCell.scala:547)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
        at akka.dispatch.Mailbox.run(Mailbox.scala:231)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
        at java.util.concurrent.ForkJoinTask.doExec(java.base@11.0.8/ForkJoinTask.java:290)
        at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(java.base@11.0.8/ForkJoinPool.java:1020)
        at java.util.concurrent.ForkJoinPool.scan(java.base@11.0.8/ForkJoinPool.java:1656)
        at java.util.concurrent.ForkJoinPool.runWorker(java.base@11.0.8/ForkJoinPool.java:1594)
        at java.util.concurrent.ForkJoinWorkerThread.run(java.base@11.0.8/ForkJoinWorkerThread.java:183)

```

and another one:

```
"application-akka.actor.default-dispatcher-11" #53 prio=5 os_prio=0 cpu=92535619.77ms elapsed=104543.86s tid=0x00007fa74c002000 nid=0x1a1b runnable  [0x00007fa77849c000]
   java.lang.Thread.State: RUNNABLE
        at akka.stream.impl.io.TLSActor.doUnwrap(TLSActor.scala:394)
        at akka.stream.impl.io.TLSActor.doInbound(TLSActor.scala:297)
        at akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1(TLSActor.scala:232)
        at akka.stream.impl.io.TLSActor$$Lambda$2467/0x0000000840ee7840.apply$mcV$sp(Unknown Source)
        at akka.stream.impl.Pump.pump(Transfer.scala:203)
        at akka.stream.impl.Pump.pump$(Transfer.scala:201)
        at akka.stream.impl.io.TLSActor.pump(TLSActor.scala:52)
        at akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse(ActorProcessor.scala:244)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
        at akka.stream.impl.SubReceive.apply(Transfer.scala:19)
        at akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse(FanOut.scala:242)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
        at akka.stream.impl.SubReceive.apply(Transfer.scala:19)
        at akka.stream.impl.SubReceive.apply(Transfer.scala:15)
        at scala.PartialFunction.applyOrElse(PartialFunction.scala:189)
        at scala.PartialFunction.applyOrElse$(PartialFunction.scala:188)
        at akka.stream.impl.SubReceive.applyOrElse(Transfer.scala:15)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:244)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:245)
        at akka.actor.Actor.aroundReceive(Actor.scala:537)
        at akka.actor.Actor.aroundReceive$(Actor.scala:535)
        at akka.stream.impl.io.TLSActor.aroundReceive(TLSActor.scala:52)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:577)
        at akka.actor.ActorCell.invoke(ActorCell.scala:547)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
        at akka.dispatch.Mailbox.run(Mailbox.scala:231)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
        at java.util.concurrent.ForkJoinTask.doExec(java.base@11.0.8/ForkJoinTask.java:290)
        at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(java.base@11.0.8/ForkJoinPool.java:1020)
        at java.util.concurrent.ForkJoinPool.scan(java.base@11.0.8/ForkJoinPool.java:1656)
        at java.util.concurrent.ForkJoinPool.runWorker(java.base@11.0.8/ForkJoinPool.java:1594)
        at java.util.concurrent.ForkJoinWorkerThread.run(java.base@11.0.8/ForkJoinWorkerThread.java:183)
```

I do not know the way to reproduce this problem. It happens sometimes after we run load test cycle on our application.

Initially we saw this problem on Akka 2.6.8 with Akka HTTP 10.1.12 running on JDK 11.0.6 on Centos 7. We updated our application and now we see the same problem on Akka 2.6.10 and Akka HTTP 10.2.2, running on JDK 11.0.8 on CentOS 7.9.

Application is REST endpoint with client TLS authentication.

**Comment by maxcom** [comment]
Maybe related to #29111

**Comment by raboof** [comment]
Thanks, that doesn't sound right indeed! Good to know that Client authentication was in use. Do you know anything about the connection that triggered this? Was it TLS1.3 or some other version, do you know what kind of client software was making the connection, etc?

**Comment by maxcom** [comment]
We do not limit TLS version, so I think that it is TLS 1.3. Keys are RSA. Client applications run on the same version of JVM. One client uses Aleph HTTP client, other uses play-ws-standalone.

Currently the only way to reproduce this case is to run 10 hour long complex performance test on our application. I'll try to find simpler reproducer.

**Comment by maxcom** [comment]
Full thread dump (Akka 2.6.8 on JDK 11.0.6).

[abook-jstack.log](https://github.com/akka/akka/files/5753832/abook-jstack.log)

**Comment by patriknw** [comment]
Could it be same as #28092 ?

**Comment by raboof** [comment]
> Could it be same as #28092 ?

Possibly - we expected that to be fixed by #29121 (released in 2.6.6) but perhaps we missed a case

**Comment by maxcom** [comment]
I found that actually we use TLS v1.2.

TLS v1.3 is disabled by default in Akka and Play.

**Comment by maxcom** [comment]
I made a reproducer for this (or similar) problem. Code sets up Akka HTTP Server and then bombs it with requests with specially broken SSLEngine. After client finishes its broken requests, server continues to eat CPU for indefinite time.

Code is here: https://github.com/maxcom/akka-http-tls-stress (use JDK 11 or later to run). Run with `sbt run`.

I reproduced problem on MacOS Mojave with AdoptOpenJDK (build 11.0.9.1+1) and CentOS 7 with OpenJDK version "11.0.10" 2021-01-19 LTS. Akka 2.6.12, Akka-HTTP 10.2.3

**Comment by jrudolph** [comment]
Thanks a lot, @maxim, I can reproduce the issue with your project. Here's another stack trace I gathered during the looping condition:

```
"server-akka.actor.default-dispatcher-21" #67 prio=5 os_prio=0 cpu=127637.79ms elapsed=149.42s tid=0x00007f6c2c001000 nid=0x1a05 runnable  [0x00007f6c7e33e000]
   java.lang.Thread.State: RUNNABLE
	at akka.stream.impl.TransferState$$anon$5.isReady(Transfer.scala:105)
	at akka.stream.impl.TransferState$$anon$6.isReady(Transfer.scala:110)
	at akka.stream.impl.io.TLSActor.doOutbound(TLSActor.scala:322)
	at akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1(TLSActor.scala:235)
	at akka.stream.impl.io.TLSActor$$Lambda$572/0x0000000800523840.apply$mcV$sp(Unknown Source)
	at akka.stream.impl.Pump.pump(Transfer.scala:203)
	at akka.stream.impl.Pump.pump$(Transfer.scala:201)
	at akka.stream.impl.io.TLSActor.pump(TLSActor.scala:52)
	at akka.stream.impl.io.TLSActor.fail(TLSActor.scala:463)
	at akka.stream.impl.io.TLSActor.doInbound(TLSActor.scala:302)
	at akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1(TLSActor.scala:232)
	at akka.stream.impl.io.TLSActor$$Lambda$572/0x0000000800523840.apply$mcV$sp(Unknown Source)
	at akka.stream.impl.Pump.pump(Transfer.scala:203)
	at akka.stream.impl.Pump.pump$(Transfer.scala:201)
	at akka.stream.impl.io.TLSActor.pump(TLSActor.scala:52)
    [...]
```

**Comment by jrudolph** [comment]
> > Could it be same as #28092 ?
> 
> Possibly - we expected that to be fixed by #29121 (released in 2.6.6) but perhaps we missed a case

This time it seems to be about error handling while in #29121 it was about a different order of things happening wrt the TLS 1.3 handshake.

**Comment by jrudolph** [comment]
For me this seems to be a JDK bug where the SslEngine (after a previous failure) can get into a state where it produces no output but still reports `NEED_WRAP` as the handshake status. I.e. it requests looping without making any progress.

**Comment by jrudolph** [comment]
I stepped through the JDK 11.0.10 code and found this:

 * engine.getHandshakeStatus() returns `NEED_WRAP` if `!outputRecord.isEmpty` (i.e. there's output ready to be sent, that's the case here)
 * on the other hand `outputRecord.isClosed` is also true, which prevents the SSLEngine from actually producing any further output

There seems to have been a similar problem before: https://bugs.openjdk.java.net/browse/JDK-8240071. It seems the original bug this one links to as duplicate is still marked as private so I don't really know about the outcome.

**Comment by jrudolph** [comment]
It also happens with JDK 8. In JDK 8, throwing away the first 2 packets arriving at the client seems to reliably trigger the behavior.

**Comment by maxcom** [comment]
Maybe introduce a handshake timeout? At least it will not spin forever.

**Comment by patriknw** [comment]
I'm reopening this because after merging PR https://github.com/akka/akka/pull/30043 the assertion triggered on Jenkins JDK11. Is that because we are using a too old JDK11 or do we have to rethink this?

https://jenkins.akka.io:8498/job/akka-nightly-jdk11/903/
```
--> [SslTls must work for TLSv1.3 must work in mode ClientInitiates while sending CompletedImmediately] Start of log messages of test that [Failed(java.lang.IllegalStateException: SSLEngine trying to loop NEED_WRAP without producing output)]
| [DEBUG] [03/10/2021 11:25:17.585] [pool-1-thread-1-ScalaTest-running-TlsSpec] [WithLogCapturing(akka://TlsSpec)] Logging started for test [SslTls must work for TLSv1.3 must work in mode ClientInitiates while sending CompletedImmediately]
| [DEBUG] [03/10/2021 11:25:17.616] [TlsSpec-akka.test.stream-dispatcher-9] [akka://TlsSpec/system/Materializers/StreamSupervisor-1/TLS-for-flow-134-2] closing output
| [DEBUG] [03/10/2021 11:25:17.616] [TlsSpec-akka.test.stream-dispatcher-17] [akka://TlsSpec/system/Materializers/StreamSupervisor-1/TLS-for-flow-134-1] closing output
| [DEBUG] [03/10/2021 11:25:17.618] [pool-1-thread-1-ScalaTest-running-TlsSpec] [WithLogCapturing(akka://TlsSpec)] Logging finished for test [SslTls must work for TLSv1.3 must work in mode ClientInitiates while sending CompletedImmediately]
<-- [SslTls must work for TLSv1.3 must work in mode ClientInitiates while sending CompletedImmediately] End of log messages of test that [Failed(java.lang.IllegalStateException: SSLEngine trying to loop NEED_WRAP without producing output)]
[info]   - must work in mode ClientInitiates while sending CompletedImmediately *** FAILED *** (45 milliseconds)
[info]     java.lang.IllegalStateException: SSLEngine trying to loop NEED_WRAP without producing output
[info]     at akka.stream.impl.io.TLSActor.doWrap(TLSActor.scala:383)
[info]     at akka.stream.impl.io.TLSActor.doOutbound(TLSActor.scala:324)
[info]     at akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1(TLSActor.scala:235)
[info]     at akka.stream.impl.Pump.pump(Transfer.scala:203)
[info]     at akka.stream.impl.Pump.pump$(Transfer.scala:201)
[info]     at akka.stream.impl.io.TLSActor.pump(TLSActor.scala:52)
[info]     at akka.stream.impl.BatchingInputBuffer.enqueueInputElement(ActorProcessor.scala:97)
[info]     at akka.stream.impl.BatchingInputBuffer$$anonfun$upstreamRunning$1.applyOrElse(ActorProcessor.scala:148)
[info]     at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
[info]     at akka.stream.impl.SubReceive.apply(Transfer.scala:19)
[info]     at akka.stream.impl.FanIn$InputBunch$$anonfun$subreceive$1.applyOrElse(FanIn.scala:244)
[info]     at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
[info]     at akka.stream.impl.SubReceive.apply(Transfer.scala:19)
[info]     at akka.stream.impl.SubReceive.apply(Transfer.scala:15)
[info]     at scala.PartialFunction.applyOrElse(PartialFunction.scala:127)
[info]     at scala.PartialFunction.applyOrElse$(PartialFunction.scala:126)
[info]     at akka.stream.impl.SubReceive.applyOrElse(Transfer.scala:15)
[info]     at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)
[info]     at akka.actor.Actor.aroundReceive(Actor.scala:537)
[info]     at akka.actor.Actor.aroundReceive$(Actor.scala:535)
[info]     at akka.stream.impl.io.TLSActor.aroundReceive(TLSActor.scala:52)
[info]     at akka.actor.ActorCell.receiveMessage(ActorCell.scala:577)
[info]     at akka.actor.ActorCell.invoke(ActorCell.scala:547)
[info]     at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
[info]     at akka.dispatch.Mailbox.run(Mailbox.scala:231)
[info]     at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
[info]     at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
[info]     at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
[info]     at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
[info]     at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
[info]     at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
[info]   - must work in mode ClientInitiates while sending CancellingRHS (559 milliseconds)
```

**Comment by jrudolph** [comment]
Also filed at #30106. It doesn't seem to fail at my machine. I see if I can reproduce it somehow.

**Comment by jrudolph** [comment]
I found [this](https://github.com/AdoptOpenJDK/openjdk-jdk11u/commit/8d1b63a4db2c6348a97b3cf45bd4d2caa7cad6b5#diff-92fd40b83d43217042269d1a80fdc42abee7430318070b903568e44e379b1197) related bug fix that came in with 11.0.8. Jenkins runs 11.0.3.

I could reproduce that the failure doesn't happen any more starting with JDK 11.0.8.

What should we do? I guess it would be fine to ignore the problem itself since it seems to be happening because of a JDK bug. The failing scenarios are all closing scenarios so they might not have any practical consequence even on older JDKs.

Should we update JDK on jenkins instead? That might not be trivial itself.

**Comment by patriknw** [comment]
I wouldn't mind that we updated the version on Jenkins, but if the original problem is fixed in latest JDK 8 and 11 versions we should maybe just revert and move on?

If we keep the check we should include JDK update hint in error message.

**Comment by maxcom** [comment]
@patriknw current issue is not fixed in latest JDK 11, infinite loop happens in 11.0.6 - 11.0.10.

I made workaround for this problem by implementing custom `SSLEngine` with handshake timeout. It spins in infinite loop up to a minute, and then throws `TimeoutException`. Not a good fix, but works.

**Comment by jrudolph** [comment]
> I wouldn't mind that we updated the version on Jenkins, but if the original problem is fixed in latest JDK 8 and 11 versions we should maybe just revert and move on?

No, there seems to be more than one issue. The failure we now see only happens on JDK < 11.0.8, i.e. there has been a similar bug in those JDK versions that we can run into (but maybe not leading to spin loops in that case).

There still is a similar but different bug reported in this ticket that is still there in latest versions of JDK 8 and JDK 11 that will lead to the spinning behavior without the PR (I filed a ticket for the JDK for that).

> If we keep the check we should include JDK update hint in error message.

The old bug happened during half-closed TLS connections, the new one during a fatal abort of the connection. Both conditions happen only happen at the end of the connection, so throwing the error guards us from spinning but is unlikely to really affect users. As far as we currently know the only side-effect of the exception will be that they get an exception from our code when they should get one from the SSL implementation anyway.

**Comment by jrudolph** [comment]
> I made workaround for this problem by implementing custom `SSLEngine` with handshake timeout. It spins in infinite loop up to a minute, and then throws `TimeoutException`. Not a good fix, but works.

I guess if you want to be more sure that nothing breaks, instead of a timeout you could also just count spins and abort after a small number of spinning rounds (while the current fix is to abort after round 1 directly). The reasoning would be that during the spin loop the SSLEngine gets no new input, so if there would be legitimate use cases of this kind of recursion, the SSLEngine code would have to converge to something after a few rounds without any more new input (which is theoretically possible but not likely).

**Comment by MaggieLeber** [comment]
This is happening to me on an otherwise running session on 2.6.15.

**Comment by istreeter** [comment]
I am seeing the problem discussed above, and I think this is new information.

Our app uses akka-http 10.2.7, and akka-stream 2.6.16, which includes the attempted fix  akka/akka#30043.  So I think the merged PR did not fully solve the problem.

We use JDK 11:

```
openjdk 11.0.11 2021-04-20
OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)
```

We have seen the same problem with JDK 17.

The symptoms are exactly the same as before: our CPU usage goes through the roof and it stops serving requests.  Even when we isolate the instance so it is no longer receiving requests, we still see CPU at 100%.

Our application runs inside docker on a EC2 instance behind a AWS load balancer.  The EC2 instance runs Ubuntu 18.04, and the docker image is Ubuntu 20.04.  AWS ALB -> EC2 node via Target Group -> TLS Listener -> TLS Termination on the Server.

We suspect the ALB is doing something unusual to trigger this meltdown.  Because we see it hit all instances in the auto-scaling group at the same time.  One minute they are all running OK and the next minute they are all stuck in the infinite loop.  We don't think it's a DOS attack because the client's TLS is terminated at the ALB, and we re-encrypt between the ALB and EC2.

Here are a couple of thread dumps that demonstrate the problem:

```
"scala-stream-collector-akka.actor.default-dispatcher-6" #22 prio=5 os_prio=0 cpu=222840.03ms elapsed=486.06s tid=0x00007fc4c8007000 nid=0x4c runnable  [0x00007fc51038f000]
   java.lang.Thread.State: RUNNABLE
        at akka.util.ByteString.isEmpty(ByteString.scala:724)
        at akka.stream.impl.io.TLSActor$ChoppingBlock.chopInto(TLSActor.scala:98)
        at akka.stream.impl.io.TLSActor.doInbound(TLSActor.scala:296)
        at akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1(TLSActor.scala:233)
        at akka.stream.impl.io.TLSActor$$Lambda$1187/0x00000008008cf840.apply$mcV$sp(Unknown Source)
        at akka.stream.impl.Pump.pump(Transfer.scala:203)
        at akka.stream.impl.Pump.pump$(Transfer.scala:201)
        at akka.stream.impl.io.TLSActor.pump(TLSActor.scala:52)
        at akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse(ActorProcessor.scala:244)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
        at akka.stream.impl.SubReceive.apply(Transfer.scala:19)
        at akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse(FanOut.scala:242)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
        at akka.stream.impl.SubReceive.apply(Transfer.scala:19)
        at akka.stream.impl.SubReceive.apply(Transfer.scala:15)
        at scala.PartialFunction.applyOrElse(PartialFunction.scala:127)
        at scala.PartialFunction.applyOrElse$(PartialFunction.scala:126)
        at akka.stream.impl.SubReceive.applyOrElse(Transfer.scala:15)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:176)
        at akka.actor.Actor.aroundReceive(Actor.scala:537)
        at akka.actor.Actor.aroundReceive$(Actor.scala:535)
        at akka.stream.impl.io.TLSActor.aroundReceive(TLSActor.scala:52)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:580)
        at akka.actor.ActorCell.invoke(ActorCell.scala:548)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
        at akka.dispatch.Mailbox.run(Mailbox.scala:231)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
        at java.util.concurrent.ForkJoinTask.doExec(java.base@11.0.11/Unknown Source)
        at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(java.base@11.0.11/Unknown Source)
        at java.util.concurrent.ForkJoinPool.scan(java.base@11.0.11/Unknown Source)
        at java.util.concurrent.ForkJoinPool.runWorker(java.base@11.0.11/Unknown Source)
        at java.util.concurrent.ForkJoinWorkerThread.run(java.base@11.0.11/Unknown Source)
```

```
"scala-stream-collector-akka.actor.default-dispatcher-14" #41 prio=5 os_prio=0 cpu=28618247.81ms elapsed=28655.32s tid=0x00007f9850217800 nid=0x65 runnable  [0x00007f98462ef000]
   java.lang.Thread.State: RUNNABLE
        at sun.security.ssl.SSLEngineImpl.unwrap(java.base@11.0.11/Unknown Source)
        - eliminated <0x0000000600eb4cf8> (a sun.security.ssl.SSLEngineImpl)
        at sun.security.ssl.SSLEngineImpl.unwrap(java.base@11.0.11/Unknown Source)
        - locked <0x0000000600eb4cf8> (a sun.security.ssl.SSLEngineImpl)
        at javax.net.ssl.SSLEngine.unwrap(java.base@11.0.11/Unknown Source)
        at akka.stream.impl.io.TLSActor.doUnwrap(TLSActor.scala:398)
        at akka.stream.impl.io.TLSActor.doInbound(TLSActor.scala:298)
        at akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1(TLSActor.scala:233)
        at akka.stream.impl.io.TLSActor$$Lambda$1192/0x00000008008cbc40.apply$mcV$sp(Unknown Source)
        at akka.stream.impl.Pump.pump(Transfer.scala:203)
        at akka.stream.impl.Pump.pump$(Transfer.scala:201)
        at akka.stream.impl.io.TLSActor.pump(TLSActor.scala:52)
        at akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse(ActorProcessor.scala:244)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
        at akka.stream.impl.SubReceive.apply(Transfer.scala:19)
        at akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse(FanOut.scala:242)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
        at akka.stream.impl.SubReceive.apply(Transfer.scala:19)
        at akka.stream.impl.SubReceive.apply(Transfer.scala:15)
        at scala.PartialFunction.applyOrElse(PartialFunction.scala:127)
        at scala.PartialFunction.applyOrElse$(PartialFunction.scala:126)
        at akka.stream.impl.SubReceive.applyOrElse(Transfer.scala:15)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:176)
        at akka.actor.Actor.aroundReceive(Actor.scala:537)
        at akka.actor.Actor.aroundReceive$(Actor.scala:535)
        at akka.stream.impl.io.TLSActor.aroundReceive(TLSActor.scala:52)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:580)
        at akka.actor.ActorCell.invoke(ActorCell.scala:548)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
        at akka.dispatch.Mailbox.run(Mailbox.scala:231)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
        at java.util.concurrent.ForkJoinTask.doExec(java.base@11.0.11/Unknown Source)
        at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(java.base@11.0.11/Unknown Source)
        at java.util.concurrent.ForkJoinPool.scan(java.base@11.0.11/Unknown Source)
        at java.util.concurrent.ForkJoinPool.runWorker(java.base@11.0.11/Unknown Source)
        at java.util.concurrent.ForkJoinWorkerThread.run(java.base@11.0.11/Unknown Source)
```

**Comment by maxcom** [comment]
@istreeter this fix actually works fine for us: https://github.com/akka/akka/issues/29922#issuecomment-796811678. I think Akka should implement handshake timeout to fix problems like this.

**Comment by dimap-tr** [comment]
@maxcom can you please share more details about this fix you implemented?

**Comment by jrudolph** [comment]
[2.6.19+37-4aeafc32-SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/com/typesafe/akka/akka-stream_2.13/2.6.19+37-4aeafc32-SNAPSHOT/) should have the potential improvement from #31321 in case someone wants to try.

**Comment by dimap-tr** [comment]
@jrudolph do you have any insight on how we can attempt and reproduce the unwrap scenario?

**Comment by tolomaus** [comment]
Hi @jrudolph,

I tried your patch (on scala 2.12 and java 11.0.15) but I still get the cpu usage of 100% by an akka thread, see here the relevant jstack output:

```
"application-akka.actor.default-dispatcher-90" #862 prio=5 os_prio=0 cpu=664752.41ms elapsed=39043.34s tid=0x00007f456c02d800 nid=0x1638 runnable  [0x00007f45663e5000]
   java.lang.Thread.State: RUNNABLE
	at sun.security.ssl.SSLEngineImpl.unwrap(java.base@11.0.15/SSLEngineImpl.java:454)
	- eliminated <0x000000009ff45638> (a sun.security.ssl.SSLEngineImpl)
	at sun.security.ssl.SSLEngineImpl.unwrap(java.base@11.0.15/SSLEngineImpl.java:433)
	- locked <0x000000009ff45638> (a sun.security.ssl.SSLEngineImpl)
	at javax.net.ssl.SSLEngine.unwrap(java.base@11.0.15/SSLEngine.java:637)
	at akka.stream.impl.io.TLSActor.doUnwrap(TLSActor.scala:399)
	at akka.stream.impl.io.TLSActor.doInbound(TLSActor.scala:298)
	at akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1(TLSActor.scala:233)
	at akka.stream.impl.io.TLSActor$$Lambda$2036/0x0000000840e16040.apply$mcV$sp(Unknown Source)
	at akka.stream.impl.Pump.pump(Transfer.scala:203)
	at akka.stream.impl.Pump.pump$(Transfer.scala:201)
	at akka.stream.impl.io.TLSActor.pump(TLSActor.scala:52)
	at akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse(ActorProcessor.scala:244)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
	at akka.stream.impl.SubReceive.apply(Transfer.scala:19)
	at akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse(FanOut.scala:242)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
	at akka.stream.impl.SubReceive.apply(Transfer.scala:19)
	at akka.stream.impl.SubReceive.apply(Transfer.scala:15)
	at scala.PartialFunction.applyOrElse(PartialFunction.scala:127)
	at scala.PartialFunction.applyOrElse$(PartialFunction.scala:126)
	at akka.stream.impl.SubReceive.applyOrElse(Transfer.scala:15)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:176)
	at akka.actor.Actor.aroundReceive(Actor.scala:537)
	at akka.actor.Actor.aroundReceive$(Actor.scala:535)
	at akka.stream.impl.io.TLSActor.aroundReceive(TLSActor.scala:52)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:580)
	at akka.actor.ActorCell.invoke(ActorCell.scala:548)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
	at akka.dispatch.Mailbox.run(Mailbox.scala:231)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
	at java.util.concurrent.ForkJoinTask.doExec(java.base@11.0.15/ForkJoinTask.java:290)
	at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(java.base@11.0.15/ForkJoinPool.java:1020)
	at java.util.concurrent.ForkJoinPool.scan(java.base@11.0.15/ForkJoinPool.java:1656)
	at java.util.concurrent.ForkJoinPool.runWorker(java.base@11.0.15/ForkJoinPool.java:1594)
	at java.util.concurrent.ForkJoinWorkerThread.run(java.base@11.0.15/ForkJoinWorkerThread.java:183)
```

Is there anything more I can do to help you fix or work around this issue?

**Comment by johanandren** [comment]
https://github.com/akka/akka/actions/runs/3473290370/jobs/5805142805#step:5:7550

Looks like this is failing often or even consistently, possibly caused by the latest CI JDK bumps.

**Comment by johanandren** [comment]
Failures were with OpenJDK 11+28 but then we bumped to Temurin 11.0.17_8 and the error stops happening.

**Comment by nirslsk** [comment]
I think this may be related, I'm running a self-contained Play 2.8 app on an EC2 instance with Amazon Linux 2 and OpenJDK 11.0.16. Every few days to a couple of weeks CPU consumption goes to 100% and stays there until I reboot the instance. The profiler I'm using shows these stack traces as the top consumers:

```
--- 23403782203 ns (36.77%), 2340 samples
  [ 0] akka.stream.impl.io.TLSActor$ChoppingBlock.putBack
  [ 1] akka.stream.impl.io.TLSActor.doUnwrap
  [ 2] akka.stream.impl.io.TLSActor.doInbound
  [ 3] akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1
  [ 4] akka.stream.impl.io.TLSActor$$Lambda$983/1363193840.apply$mcV$sp
  [ 5] akka.stream.impl.Pump.pump
  [ 6] akka.stream.impl.Pump.pump$
  [ 7] akka.stream.impl.io.TLSActor.pump
  [ 8] akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse
  [ 9] scala.runtime.AbstractPartialFunction.apply
  [10] akka.stream.impl.SubReceive.apply
  [11] akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse
  [12] scala.runtime.AbstractPartialFunction.apply
  [13] akka.stream.impl.SubReceive.apply
  [14] akka.stream.impl.SubReceive.apply
  [15] scala.PartialFunction.applyOrElse
  [16] scala.PartialFunction.applyOrElse$
  [17] akka.stream.impl.SubReceive.applyOrElse
  [18] scala.PartialFunction$OrElse.applyOrElse
  [19] scala.PartialFunction$OrElse.applyOrElse
  [20] akka.actor.Actor.aroundReceive
  [21] akka.actor.Actor.aroundReceive$
  [22] akka.stream.impl.io.TLSActor.aroundReceive
  [23] akka.actor.ActorCell.receiveMessage
  [24] akka.actor.ActorCell.invoke
  [25] akka.dispatch.Mailbox.processMailbox
  [26] akka.dispatch.Mailbox.run
  [27] akka.dispatch.Mailbox.exec
  [28] java.util.concurrent.ForkJoinTask.doExec
  [29] java.util.concurrent.ForkJoinPool$WorkQueue.runTask
  [30] java.util.concurrent.ForkJoinPool.runWorker
  [31] java.util.concurrent.ForkJoinWorkerThread.run

--- 15005141980 ns (23.58%), 1500 samples
  [ 0] jbyte_disjoint_arraycopy
  [ 1] akka.stream.impl.io.TLSActor$ChoppingBlock.putBack
  [ 2] akka.stream.impl.io.TLSActor.doUnwrap
  [ 3] akka.stream.impl.io.TLSActor.doInbound
  [ 4] akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1
  [ 5] akka.stream.impl.io.TLSActor$$Lambda$983/1363193840.apply$mcV$sp
  [ 6] akka.stream.impl.Pump.pump
  [ 7] akka.stream.impl.Pump.pump$
  [ 8] akka.stream.impl.io.TLSActor.pump
  [ 9] akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse
  [10] scala.runtime.AbstractPartialFunction.apply
  [11] akka.stream.impl.SubReceive.apply
  [12] akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse
  [13] scala.runtime.AbstractPartialFunction.apply
  [14] akka.stream.impl.SubReceive.apply
  [15] akka.stream.impl.SubReceive.apply
  [16] scala.PartialFunction.applyOrElse
  [17] scala.PartialFunction.applyOrElse$
  [18] akka.stream.impl.SubReceive.applyOrElse
  [19] scala.PartialFunction$OrElse.applyOrElse
  [20] scala.PartialFunction$OrElse.applyOrElse
  [21] akka.actor.Actor.aroundReceive
  [22] akka.actor.Actor.aroundReceive$
  [23] akka.stream.impl.io.TLSActor.aroundReceive
  [24] akka.actor.ActorCell.receiveMessage
  [25] akka.actor.ActorCell.invoke
  [26] akka.dispatch.Mailbox.processMailbox
  [27] akka.dispatch.Mailbox.run
  [28] akka.dispatch.Mailbox.exec
  [29] java.util.concurrent.ForkJoinTask.doExec
  [30] java.util.concurrent.ForkJoinPool$WorkQueue.runTask
  [31] java.util.concurrent.ForkJoinPool.runWorker
  [32] java.util.concurrent.ForkJoinWorkerThread.run

--- 8741553485 ns (13.74%), 874 samples
  [ 0] jbyte_disjoint_arraycopy
  [ 1] akka.util.ByteString$ByteString1C.writeToBuffer
  [ 2] akka.util.ByteString$ByteString1C.copyToBuffer
  [ 3] akka.stream.impl.io.TLSActor$ChoppingBlock.chopInto
  [ 4] akka.stream.impl.io.TLSActor.doInbound
  [ 5] akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1
  [ 6] akka.stream.impl.io.TLSActor$$Lambda$983/1363193840.apply$mcV$sp
  [ 7] akka.stream.impl.Pump.pump
  [ 8] akka.stream.impl.Pump.pump$
  [ 9] akka.stream.impl.io.TLSActor.pump
  [10] akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse
  [11] scala.runtime.AbstractPartialFunction.apply
  [12] akka.stream.impl.SubReceive.apply
  [13] akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse
  [14] scala.runtime.AbstractPartialFunction.apply
  [15] akka.stream.impl.SubReceive.apply
  [16] akka.stream.impl.SubReceive.apply
  [17] scala.PartialFunction.applyOrElse
  [18] scala.PartialFunction.applyOrElse$
  [19] akka.stream.impl.SubReceive.applyOrElse
  [20] scala.PartialFunction$OrElse.applyOrElse
  [21] scala.PartialFunction$OrElse.applyOrElse
  [22] akka.actor.Actor.aroundReceive
  [23] akka.actor.Actor.aroundReceive$
  [24] akka.stream.impl.io.TLSActor.aroundReceive
  [25] akka.actor.ActorCell.receiveMessage
  [26] akka.actor.ActorCell.invoke
  [27] akka.dispatch.Mailbox.processMailbox
  [28] akka.dispatch.Mailbox.run
  [29] akka.dispatch.Mailbox.exec
  [30] java.util.concurrent.ForkJoinTask.doExec
  [31] java.util.concurrent.ForkJoinPool$WorkQueue.runTask
  [32] java.util.concurrent.ForkJoinPool.runWorker
  [33] java.util.concurrent.ForkJoinWorkerThread.run
```

**Comment by saurabh47** [comment]
we are also facing exact same issue mentioned by @nirslsk 
Any solution ?
> I think this may be related, I'm running a self-contained Play 2.8.19 app on an EC2 instance with Amazon Linux 2 and OpenJDK 11.0.16. Every few days to a couple of weeks CPU consumption goes to 100% and stays there until I reboot the instance. The profiler I'm using shows these stack traces as the top consumers:
> 
> ```
> --- 23403782203 ns (36.77%), 2340 samples
>   [ 0] akka.stream.impl.io.TLSActor$ChoppingBlock.putBack
>   [ 1] akka.stream.impl.io.TLSActor.doUnwrap
>   [ 2] akka.stream.impl.io.TLSActor.doInbound
>   [ 3] akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1
>   [ 4] akka.stream.impl.io.TLSActor$$Lambda$983/1363193840.apply$mcV$sp
>   [ 5] akka.stream.impl.Pump.pump
>   [ 6] akka.stream.impl.Pump.pump$
>   [ 7] akka.stream.impl.io.TLSActor.pump
>   [ 8] akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse
>   [ 9] scala.runtime.AbstractPartialFunction.apply
>   [10] akka.stream.impl.SubReceive.apply
>   [11] akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse
>   [12] scala.runtime.AbstractPartialFunction.apply
>   [13] akka.stream.impl.SubReceive.apply
>   [14] akka.stream.impl.SubReceive.apply
>   [15] scala.PartialFunction.applyOrElse
>   [16] scala.PartialFunction.applyOrElse$
>   [17] akka.stream.impl.SubReceive.applyOrElse
>   [18] scala.PartialFunction$OrElse.applyOrElse
>   [19] scala.PartialFunction$OrElse.applyOrElse
>   [20] akka.actor.Actor.aroundReceive
>   [21] akka.actor.Actor.aroundReceive$
>   [22] akka.stream.impl.io.TLSActor.aroundReceive
>   [23] akka.actor.ActorCell.receiveMessage
>   [24] akka.actor.ActorCell.invoke
>   [25] akka.dispatch.Mailbox.processMailbox
>   [26] akka.dispatch.Mailbox.run
>   [27] akka.dispatch.Mailbox.exec
>   [28] java.util.concurrent.ForkJoinTask.doExec
>   [29] java.util.concurrent.ForkJoinPool$WorkQueue.runTask
>   [30] java.util.concurrent.ForkJoinPool.runWorker
>   [31] java.util.concurrent.ForkJoinWorkerThread.run
> 
> --- 15005141980 ns (23.58%), 1500 samples
>   [ 0] jbyte_disjoint_arraycopy
>   [ 1] akka.stream.impl.io.TLSActor$ChoppingBlock.putBack
>   [ 2] akka.stream.impl.io.TLSActor.doUnwrap
>   [ 3] akka.stream.impl.io.TLSActor.doInbound
>   [ 4] akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1
>   [ 5] akka.stream.impl.io.TLSActor$$Lambda$983/1363193840.apply$mcV$sp
>   [ 6] akka.stream.impl.Pump.pump
>   [ 7] akka.stream.impl.Pump.pump$
>   [ 8] akka.stream.impl.io.TLSActor.pump
>   [ 9] akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse
>   [10] scala.runtime.AbstractPartialFunction.apply
>   [11] akka.stream.impl.SubReceive.apply
>   [12] akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse
>   [13] scala.runtime.AbstractPartialFunction.apply
>   [14] akka.stream.impl.SubReceive.apply
>   [15] akka.stream.impl.SubReceive.apply
>   [16] scala.PartialFunction.applyOrElse
>   [17] scala.PartialFunction.applyOrElse$
>   [18] akka.stream.impl.SubReceive.applyOrElse
>   [19] scala.PartialFunction$OrElse.applyOrElse
>   [20] scala.PartialFunction$OrElse.applyOrElse
>   [21] akka.actor.Actor.aroundReceive
>   [22] akka.actor.Actor.aroundReceive$
>   [23] akka.stream.impl.io.TLSActor.aroundReceive
>   [24] akka.actor.ActorCell.receiveMessage
>   [25] akka.actor.ActorCell.invoke
>   [26] akka.dispatch.Mailbox.processMailbox
>   [27] akka.dispatch.Mailbox.run
>   [28] akka.dispatch.Mailbox.exec
>   [29] java.util.concurrent.ForkJoinTask.doExec
>   [30] java.util.concurrent.ForkJoinPool$WorkQueue.runTask
>   [31] java.util.concurrent.ForkJoinPool.runWorker
>   [32] java.util.concurrent.ForkJoinWorkerThread.run
> 
> --- 8741553485 ns (13.74%), 874 samples
>   [ 0] jbyte_disjoint_arraycopy
>   [ 1] akka.util.ByteString$ByteString1C.writeToBuffer
>   [ 2] akka.util.ByteString$ByteString1C.copyToBuffer
>   [ 3] akka.stream.impl.io.TLSActor$ChoppingBlock.chopInto
>   [ 4] akka.stream.impl.io.TLSActor.doInbound
>   [ 5] akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1
>   [ 6] akka.stream.impl.io.TLSActor$$Lambda$983/1363193840.apply$mcV$sp
>   [ 7] akka.stream.impl.Pump.pump
>   [ 8] akka.stream.impl.Pump.pump$
>   [ 9] akka.stream.impl.io.TLSActor.pump
>   [10] akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse
>   [11] scala.runtime.AbstractPartialFunction.apply
>   [12] akka.stream.impl.SubReceive.apply
>   [13] akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse
>   [14] scala.runtime.AbstractPartialFunction.apply
>   [15] akka.stream.impl.SubReceive.apply
>   [16] akka.stream.impl.SubReceive.apply
>   [17] scala.PartialFunction.applyOrElse
>   [18] scala.PartialFunction.applyOrElse$
>   [19] akka.stream.impl.SubReceive.applyOrElse
>   [20] scala.PartialFunction$OrElse.applyOrElse
>   [21] scala.PartialFunction$OrElse.applyOrElse
>   [22] akka.actor.Actor.aroundReceive
>   [23] akka.actor.Actor.aroundReceive$
>   [24] akka.stream.impl.io.TLSActor.aroundReceive
>   [25] akka.actor.ActorCell.receiveMessage
>   [26] akka.actor.ActorCell.invoke
>   [27] akka.dispatch.Mailbox.processMailbox
>   [28] akka.dispatch.Mailbox.run
>   [29] akka.dispatch.Mailbox.exec
>   [30] java.util.concurrent.ForkJoinTask.doExec
>   [31] java.util.concurrent.ForkJoinPool$WorkQueue.runTask
>   [32] java.util.concurrent.ForkJoinPool.runWorker
>   [33] java.util.concurrent.ForkJoinWorkerThread.run
> ```

**Comment by patriknw** [comment]
Which JDK are you using. Also OpenJDK 11.0.16?
Would you be able to switch to Temurin 11.0.17_8 (or later), at least to confirm that it's related to the JDK vendor/version?

**Comment by nirslsk** [comment]
Yes OpenJDK 11.0.16, I'll try Temurin but it takes 1-3 weeks for the problem to surface (maybe because my server is low-traffic), so it's going to be a while before I can report with any confidence that it's not happening.

**Comment by nirslsk** [comment]
On Temurin 11.0.18+10  now. Will report back if the problem happens again or in a few weeks if it doesn't.

**Comment by patriknw** [comment]
Thanks

**Comment by saurabh47** [comment]
@patriknw / @johanandren

we are again facing a similar high CPU issue on Temurin 11.0.18+10
![image](https://user-images.githubusercontent.com/15173289/225595413-3e244fcc-9218-4a84-83ce-c17010273462.png)
### CPU Usage
![image](https://user-images.githubusercontent.com/15173289/225595554-a90675e5-ee71-4d2a-8402-fe7665de0f31.png)

I have attached the thread dump. Please take a look 
[temurin-jdk11-threaddump.txt](https://github.com/akka/akka/files/10989743/temurin-jdk11-threaddump.txt)

**Comment by johanandren** [comment]
Is that with Akka 2.6.20 or later (after https://github.com/akka/akka/pull/31321 was merged)?

**Comment by saurabh47** [comment]
hey @johanandren,
we are not directly using the Akka framework. we using the scala play framework.
we are using version of that
```
scalaVersion := "2.13.1"
```
```
addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.8.0")
```

**Comment by johanandren** [comment]
2.8.0 is relatively old and uses Akka 2.6.0 which is before that fix was merged. Try bumping to the latest Playframework 2.8.19, it should bring up Akka to 2.6.20 which contains that fix, and see if that helps.

**Comment by saurabh47** [comment]
sure will try Playframework 2.8.19 & let you know the results

**Comment by saurabh47** [comment]
@johanandren with Playframework 2.8.19 as well we facing the similar CPU utilization issue
I have attached thread dumps
Please take a look. Let me know if you need more information

[th-p1.txt](https://github.com/akka/akka/files/11001117/th-p1.txt)
[th-p2.txt](https://github.com/akka/akka/files/11001118/th-p2.txt)

```
java.lang.Thread.State: WAITING (parking)
at jdk.internal.misc.Unsafe.park(java.base@11.0.18/Native Method)
- parking to wait for  <0x0000000090e6b5b8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
at java.util.concurrent.locks.LockSupport.park(java.base@11.0.18/LockSupport.java:194)
at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@11.0.18/AbstractQueuedSynchronizer.java:2081)


java.lang.Thread.State: RUNNABLE
at sun.security.ssl.SSLEngineImpl.checkParams(java.base@11.0.18/SSLEngineImpl.java:408)
at sun.security.ssl.SSLEngineImpl.unwrap(java.base@11.0.18/SSLEngineImpl.java:451)
- locked <0x0000000093c74ba8> (a sun.security.ssl.SSLEngineImpl)
at sun.security.ssl.SSLEngineImpl.unwrap(java.base@11.0.18/SSLEngineImpl.java:433)

java.lang.Thread.State: TIMED_WAITING (parking)
at jdk.internal.misc.Unsafe.park(java.base@11.0.18/Native Method)
- parking to wait for  <0x0000000090ef9380> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
at java.util.concurrent.locks.LockSupport.parkNanos(java.base@11.0.18/LockSupport.java:234)
at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(java.base@11.0.18/AbstractQueuedSynchronizer.java:2123)

```
![image](https://user-images.githubusercontent.com/15173289/225900167-55d14dce-42e6-48ce-9760-3e95c5df1c61.png)

**Comment by PrabuRepo** [comment]
Hello,

We are also facing the similar issue in our API. Infinite loop in tls actor and it leads to high cpu utilization in few threads randomly. 

Tech Stack: openjdk8, akka-2.6.20, akka-http-10.2.7
Infrastructure: ALB + ECS with EC2 instance
TLS 1.2

Any suggestions/solutions for this issue? We have been using akka framework for more than 2 years with 2.6.12 version without any issue, but this issue started 2 months ago and based on above recommendation, updated to 2.6.20, but still issue persists. 
Any one can explain why this issue occurs, if you can share any reference or links to understand this issue will be helpful. 

Please let me if you need more details.

**Comment by johanandren** [comment]
Several reports state they are on EC2, so maybe it is something specifically triggered by something ELB does. If someone can figure out a way to consistently repeat this it would be very useful.

**Comment by krishnasaivst** [comment]
> Can you share anything more about the specific infrastructure, is it external clients hitting the TLS endpoint, is it some specific load balancer inbetween, do you have some way to repeat that consistently triggers the spin?

The pods are all running on the EKS cluster with the EC2 instances as the nodes. The pods are served externally using the ALB. Every time we scale down and scale up the pod we see the high utilization of the CPU.

**Comment by nirslsk** [comment]
Seems to still be happening... let me know if you want me to try anything or need access to the instance. I'm not using any load balancer, just a single small EC2 instance with Temurin.

```
--- 18417772939 ns (30.35%), 1842 samples
  [ 0] akka.stream.impl.io.TLSActor$ChoppingBlock.putBack
  [ 1] akka.stream.impl.io.TLSActor.doUnwrap
  [ 2] akka.stream.impl.io.TLSActor.doInbound
  [ 3] akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1
  [ 4] akka.stream.impl.io.TLSActor$$Lambda$1076/1522455354.apply$mcV$sp
  [ 5] akka.stream.impl.Pump.pump
  [ 6] akka.stream.impl.Pump.pump$
  [ 7] akka.stream.impl.io.TLSActor.pump
  [ 8] akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse
  [ 9] scala.runtime.AbstractPartialFunction.apply
  [10] akka.stream.impl.SubReceive.apply
  [11] akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse
  [12] scala.runtime.AbstractPartialFunction.apply
  [13] akka.stream.impl.SubReceive.apply
  [14] akka.stream.impl.SubReceive.apply
  [15] scala.PartialFunction.applyOrElse
  [16] scala.PartialFunction.applyOrElse$
  [17] akka.stream.impl.SubReceive.applyOrElse
  [18] scala.PartialFunction$OrElse.applyOrElse
  [19] scala.PartialFunction$OrElse.applyOrElse
  [20] akka.actor.Actor.aroundReceive
  [21] akka.actor.Actor.aroundReceive$
  [22] akka.stream.impl.io.TLSActor.aroundReceive
  [23] akka.actor.ActorCell.receiveMessage
  [24] akka.actor.ActorCell.invoke
  [25] akka.dispatch.Mailbox.processMailbox
  [26] akka.dispatch.Mailbox.run
  [27] akka.dispatch.Mailbox.exec
  [28] java.util.concurrent.ForkJoinTask.doExec
  [29] java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec
  [30] java.util.concurrent.ForkJoinPool.scan
  [31] java.util.concurrent.ForkJoinPool.runWorker
  [32] java.util.concurrent.ForkJoinWorkerThread.run

--- 12431783749 ns (20.49%), 1242 samples
  [ 0] jbyte_disjoint_arraycopy
  [ 1] akka.stream.impl.io.TLSActor$ChoppingBlock.putBack
  [ 2] akka.stream.impl.io.TLSActor.doUnwrap
  [ 3] akka.stream.impl.io.TLSActor.doInbound
  [ 4] akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1
  [ 5] akka.stream.impl.io.TLSActor$$Lambda$1076/1522455354.apply$mcV$sp
  [ 6] akka.stream.impl.Pump.pump
  [ 7] akka.stream.impl.Pump.pump$
  [ 8] akka.stream.impl.io.TLSActor.pump
  [ 9] akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse
  [10] scala.runtime.AbstractPartialFunction.apply
  [11] akka.stream.impl.SubReceive.apply
  [12] akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse
  [13] scala.runtime.AbstractPartialFunction.apply
  [14] akka.stream.impl.SubReceive.apply
  [15] akka.stream.impl.SubReceive.apply
  [16] scala.PartialFunction.applyOrElse
  [17] scala.PartialFunction.applyOrElse$
  [18] akka.stream.impl.SubReceive.applyOrElse
  [19] scala.PartialFunction$OrElse.applyOrElse
  [20] scala.PartialFunction$OrElse.applyOrElse
  [21] akka.actor.Actor.aroundReceive
  [22] akka.actor.Actor.aroundReceive$
  [23] akka.stream.impl.io.TLSActor.aroundReceive
  [24] akka.actor.ActorCell.receiveMessage
  [25] akka.actor.ActorCell.invoke
  [26] akka.dispatch.Mailbox.processMailbox
  [27] akka.dispatch.Mailbox.run
  [28] akka.dispatch.Mailbox.exec
  [29] java.util.concurrent.ForkJoinTask.doExec
  [30] java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec
  [31] java.util.concurrent.ForkJoinPool.scan
  [32] java.util.concurrent.ForkJoinPool.runWorker
  [33] java.util.concurrent.ForkJoinWorkerThread.run

--- 9872384558 ns (16.27%), 988 samples
  [ 0] __memset_avx2_erms
  [ 1] MemAllocator::allocate() const
  [ 2] CollectedHeap::array_allocate(Klass*, int, int, bool, Thread*)
  [ 3] TypeArrayKlass::allocate_common(int, bool, Thread*)
  [ 4] OptoRuntime::new_array_C(Klass*, int, JavaThread*)
  [ 5] akka.util.ByteString$ByteString1C$.apply
  [ 6] akka.util.CompactByteString$.apply
  [ 7] akka.util.ByteString$.apply
  [ 8] akka.stream.impl.io.TLSActor$ChoppingBlock.putBack
  [ 9] akka.stream.impl.io.TLSActor.doUnwrap
  [10] akka.stream.impl.io.TLSActor.doInbound
  [11] akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1
  [12] akka.stream.impl.io.TLSActor$$Lambda$1076/1522455354.apply$mcV$sp
  [13] akka.stream.impl.Pump.pump
  [14] akka.stream.impl.Pump.pump$
  [15] akka.stream.impl.io.TLSActor.pump
  [16] akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse
  [17] scala.runtime.AbstractPartialFunction.apply
  [18] akka.stream.impl.SubReceive.apply
  [19] akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse
  [20] scala.runtime.AbstractPartialFunction.apply
  [21] akka.stream.impl.SubReceive.apply
  [22] akka.stream.impl.SubReceive.apply
  [23] scala.PartialFunction.applyOrElse
  [24] scala.PartialFunction.applyOrElse$
  [25] akka.stream.impl.SubReceive.applyOrElse
  [26] scala.PartialFunction$OrElse.applyOrElse
  [27] scala.PartialFunction$OrElse.applyOrElse
  [28] akka.actor.Actor.aroundReceive
  [29] akka.actor.Actor.aroundReceive$
  [30] akka.stream.impl.io.TLSActor.aroundReceive
  [31] akka.actor.ActorCell.receiveMessage
  [32] akka.actor.ActorCell.invoke
  [33] akka.dispatch.Mailbox.processMailbox
  [34] akka.dispatch.Mailbox.run
  [35] akka.dispatch.Mailbox.exec
  [36] java.util.concurrent.ForkJoinTask.doExec
  [37] java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec
  [38] java.util.concurrent.ForkJoinPool.scan
  [39] java.util.concurrent.ForkJoinPool.runWorker
  [40] java.util.concurrent.ForkJoinWorkerThread.run

--- 7644481779 ns (12.60%), 763 samples
  [ 0] jbyte_disjoint_arraycopy
  [ 1] java.nio.ByteBuffer.put
  [ 2] java.nio.DirectByteBuffer.put
  [ 3] akka.util.ByteString$ByteString1C.writeToBuffer
  [ 4] akka.util.ByteString$ByteString1C.copyToBuffer
  [ 5] akka.stream.impl.io.TLSActor$ChoppingBlock.chopInto
  [ 6] akka.stream.impl.io.TLSActor.doInbound
  [ 7] akka.stream.impl.io.TLSActor.$anonfun$bidirectional$1
  [ 8] akka.stream.impl.io.TLSActor$$Lambda$1076/1522455354.apply$mcV$sp
  [ 9] akka.stream.impl.Pump.pump
  [10] akka.stream.impl.Pump.pump$
  [11] akka.stream.impl.io.TLSActor.pump
  [12] akka.stream.impl.SimpleOutputs$$anonfun$downstreamRunning$1.applyOrElse
  [13] scala.runtime.AbstractPartialFunction.apply
  [14] akka.stream.impl.SubReceive.apply
  [15] akka.stream.impl.FanOut$OutputBunch$$anonfun$subreceive$1.applyOrElse
  [16] scala.runtime.AbstractPartialFunction.apply
  [17] akka.stream.impl.SubReceive.apply
  [18] akka.stream.impl.SubReceive.apply
  [19] scala.PartialFunction.applyOrElse
  [20] scala.PartialFunction.applyOrElse$
  [21] akka.stream.impl.SubReceive.applyOrElse
  [22] scala.PartialFunction$OrElse.applyOrElse
  [23] scala.PartialFunction$OrElse.applyOrElse
  [24] akka.actor.Actor.aroundReceive
  [25] akka.actor.Actor.aroundReceive$
  [26] akka.stream.impl.io.TLSActor.aroundReceive
  [27] akka.actor.ActorCell.receiveMessage
  [28] akka.actor.ActorCell.invoke
  [29] akka.dispatch.Mailbox.processMailbox
  [30] akka.dispatch.Mailbox.run
  [31] akka.dispatch.Mailbox.exec
  [32] java.util.concurrent.ForkJoinTask.doExec
  [33] java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec
  [34] java.util.concurrent.ForkJoinPool.scan
  [35] java.util.concurrent.ForkJoinPool.runWorker
  [36] java.util.concurrent.ForkJoinWorkerThread.run
```

**Comment by johanandren** [comment]
If you could figure out what type of client it is whose connection triggers the infinite loop, and best case, figure out how we can reproduce it locally that would help.

**Comment by Sulbigar** [comment]
@johanandren - I'm not sure if the problem is exactly in unwrap. In our case, I took the thread dump of the long running thread (which consumes CPU) in an interval of 3-4 secs and it seems that the Transfer.pump function keeps running  running in loop. [This](https://github.com/akka/akka/blob/v2.6.20/akka-stream/src/main/scala/akka/stream/impl/Transfer.scala#L202) is where I see it happening. Not exactly sure what causes this but wanted to share this info to see if a possibility can be traced. We are behind ECS though.

Akka 2.6.20

![Untitled](https://user-images.githubusercontent.com/12649226/228035036-4c5dc035-9224-4010-8b66-081ad7a5acc0.gif)

**Comment by johanandren** [comment]
Thanks @Sulbigar that's interesting information.

**Comment by saurabh47** [comment]
@johanandren we have tried running our app on HTTP without SSL but still, it was hitting 90%+ CPU on the Kubernetes pods
I have attached a thread dump here

[jstack.txt](https://github.com/akka/akka/files/11089335/jstack.txt)

**Comment by johanandren** [comment]
@saurabh47 surprising that you see such CPU usage, thread dump looks like an idle system. Could it be a sidecar container/process consuming all that CPU perhaps?

**Comment by johanandren** [comment]
Both CPU columns are 0.0 all the way across, seems unrelated to this TLS issue and more likely something with your infra?

**Comment by nirslsk** [comment]
Thank you Johan! Took me a while to figure out how to use the snapshot in my Play app. Just in case it can help someone or someone wants to tell me I did it wrong, I added these lines to build.sbt:

```
resolvers += "Snapshots" at "https://oss.sonatype.org/content/repositories/snapshots/"

libraryDependencies ++= Seq(
  "com.typesafe.akka" % "akka-remote_2.13" % "2.8.1-M1+3-6b7b031c-SNAPSHOT",
  "com.typesafe.akka" % "akka-slf4j_2.13" % "2.8.1-M1+3-6b7b031c-SNAPSHOT",
  "com.typesafe.akka" % "akka-actor-typed_2.13" % "2.8.1-M1+3-6b7b031c-SNAPSHOT",
  "com.typesafe.akka" % "akka-protobuf-v3_2.13" % "2.8.1-M1+3-6b7b031c-SNAPSHOT",
  "com.typesafe.akka" % "akka-stream_2.13" % "2.8.1-M1+3-6b7b031c-SNAPSHOT",
  "com.typesafe.akka" % "akka-serialization-jackson_2.13" % "2.8.1-M1+3-6b7b031c-SNAPSHOT")
```

Verified at least locally that it's using the version of doUnwrap that has the extra loops parameter, and in production the 2.13-2.8.1-M1+3-6b7b031c-SNAPSHOT versions show up in the lib folder of my staged Play app so I hope everything's set up correctly. Will report in due time.

**Comment by DaveDeCaprio** [comment]
We have also tried the snapshot and are still seeing the same issue.

**Comment by johanandren** [comment]
Thanks for confirming, I have already closed the PR as non-fix for the issue.

We still lack insight into exactly what is going on, so if anybody able to create a stable reproducer or figure out something more about what is causing it will be needed before we can make any progress on a fix.

**Comment by nirslsk** [comment]
I set up the app for remote debugging, so I'm able to use IntelliJ to debug it next time it happens.

**Comment by DaveDeCaprio** [comment]
I still don't have a consistent way to reproduce, but I'v ediagnosed this a bit further.

The problem is in TLSActor.pump.  It gets stuck in a loop where the SSL connection is in the "bidirectional" state and the lastHandshakeStatus is NEEDS_WRAP.  

I'm not 100% sure on this, but what looks like is happening:
The system is pulling data using `transportInChoppingBlock.chopInto(transportInBuffer)` (line 297 of TLSActor) 
and then since the lastHandshake status is NEEDS_WRAP, it then just puts that back on 'transportInChoppingBlock.putBack(transportInBuffer)' line 412.
and then Pump.pump still thinks everything is executable (Transfer.scala line 202).

It looks like the transportInChoppingBlock having bytes and also "outputBunch.demandAvailableFor(UserOut)" should be sufficient to keep the stream executable.

In our case, this occurs long after the SSL connection is terminated from the client, so the problem might be in the "demandAvailableFor" function?

I'll also try to get a debugger on this to get more insight.  I can't replicate it outside our production environment.

**Comment by rmmeans** [comment]
Awesome research @DaveDeCaprio! I don't want to distract from that information as that is likely way more helpful than what I have to share.

We also started seeing this issue hit us at the very end of February. We run on AWS Fargate with an ALB load balancer in front of our servers which are running Play 2.8.19. A thread dump led me to this github issue and based on some other feedback and noticing the uptick in comments since March from people running on AWS, I may have found a correlation:

https://aws.amazon.com/about-aws/whats-new/2023/03/application-load-balancer-tls-1-3/

I contacted AWS Support and they confirmed that the rollout of support started in the weeks prior as a gradual rollout. I am waiting on confirmation from AWS, but I suspect that the rollout also enabled TLS 1.3 on the backend connection to our running Play containers.

Our experience has been the same as others in this thread - it appears to be random, usually for us only a single container out of 5 will get hit at time and we have to go kill that container to stop the CPU burn. You can see in the following chart all of these long-running spikes and eventually stopping the offending container. Sometimes the issue would go away for a few days, other times only for an hour or less. 

![image](https://github.com/akka/akka/assets/3507512/9d68f58f-54f0-490c-be61-636fe8df4ebb)

While I don't like suggesting this, I do want to share that we were able to mitigate this issue on Play framework by switching Play server backend to Netty, you can see that in the graph when we switched on On 05/03 and the issue has not come back. We lost some performance in terms of CPU, but our P99.99 latency is close enough and worth the stable system. I'll be watching this thread and hopefully a fix will come that Play can use and we can switch back. 👀

**Comment by nirslsk** [comment]
Dave clearly has a much better understanding of this than I do, but I have it in a debugger and what I'm seeing is doOutbound being called over and over again from line 236 of TLSActor.scala, and doOutbound failing all its if conditions over and over again, doing nothing. I'll keep it suspended in a debugger for as long as I can, let me know if there's anything you want me to try out.

<img width="1440" alt="Screen Shot 2023-05-19 at 3 03 34 AM" src="https://github.com/akka/akka/assets/2808762/0acd0ce4-e5ee-460c-90c2-562aece2dbd2">

**Comment by nirslsk** [comment]
What Dave is describing seems to be happening... the outermost block of code it's stuck on is this one in Transfers.scala, currentAction() is called repeatedly:

<img width="908" alt="Screen Shot 2023-05-19 at 3 43 20 AM" src="https://github.com/akka/akka/assets/2808762/49e1be4b-8188-41ac-80ee-60685b9ad4dd">

currentAction() seems to call this block of code in TLSActor.scala:

<img width="658" alt="Screen Shot 2023-05-19 at 3 44 13 AM" src="https://github.com/akka/akka/assets/2808762/222a98c5-3b38-4492-9bf9-eba9ff86cf3a">

doInbound() skips straight to this part, where chopInto() and doUnwrap() are called one after the other:

<img width="968" alt="Screen Shot 2023-05-19 at 3 49 34 AM" src="https://github.com/akka/akka/assets/2808762/a08d9102-a46c-46df-afe9-d472f96f871d">

doUnwrap() handles a NEED_WRAP handshake status and calls transportInChoppingBlock.putBack()

<img width="989" alt="Screen Shot 2023-05-19 at 3 51 23 AM" src="https://github.com/akka/akka/assets/2808762/91d5c055-25b1-4aac-bfa2-dada233f9ce6">

Then it's back to the second block of code where doOutbound gets called and does nothing, and back up to the first block of code, repeat endlessly.

**Comment by DaveDeCaprio** [comment]
Great.  I haven't caught it in a debugger yet.
At some point the connection is getting dropped but it isn't transitioning out of the "bidirectional" state.

**Comment by nirslsk** [comment]
Anything you want me to look into before I reboot my instance?

**Comment by johanandren** [comment]
Based on the last bits of input here I have tried another to do another work around that should bail out of the loop in #31947 

I have published a snapshot with version `2.8.1+3-7ead7d48-SNAPSHOT` of that PR to the sonatype snapshots repo, please try it out and see if it avoids the CPU hogging.

**Comment by DaveDeCaprio** [comment]
We will try this out and let you know if it addresses the issue.

 

Dave

**Comment by DaveDeCaprio** [comment]
We haven't seen this since we rolled out the new patch, and we were getting it pretty consistently before in production.  

We are closing our internal bug ticket for this.  I'd love to see this PR merged into the main line.  I know it's a bit of a hack, but it's made a huge difference for us.

**Comment by DaveDeCaprio** [comment]
It's not in our logs, but we might have logging turned down for Akka in production.

**Comment by nkanitkar** [comment]
We have tested this patch in our dev environment and it appears to have resolved the issue. I didn't see the exception logged, but most likely due to our logging levels.

Testing Results:


| Akka Version  | Recreated issue |
| ------------- | ------------- |
| 2.7.0  | Yes  |
| 2.8.2  | Yes  | 
| 2.8.1+3-7ead7d48-SNAPSHOT  | No  |

**Comment by dimap-tr** [comment]
Are there any plans to backport this fix to 2.6.x ? 2.7 and higher are released only under BUSL which is a no go for us (and I am sure for many others)

**Comment by johanandren** [comment]
We have brought this up with product management, but not decided yet.

**Comment by nirslsk** [comment]
In the meantime, switching to Netty as a backend if you use Play seems to work very well:
 https://www.playframework.com/documentation/2.8.x/NettyServer

**Comment by jtownley** [comment]
Hello watchers. we will be backporting a fix for this issue to akka 2.6.

**Comment by johanandren** [comment]
2.6.21 is now released with a back ported fix for this, the fix will also be in Akka 2.8.3 release when that is released.
